typescript-test-samples/schema-and-contract-testing/README.md

```diff
@@ -0,0 +1,109 @@
+
+Both of these limitations can be addressed with [Contract testing](#contract-testing).
+
+## Contract Testing
+
+### System Under Test (SUT)
+
+This test is integration service agnostic as it uses sample events generated from schema published to a schema registry. The diagram below shows one way of implementing contract testing as part of producer’s CI/CD pipeline.
+
+The diagram below shows an example of running the schema validation as part of a producer’s CI/CD pipeline. A sample unit test from the "Build" step is provided in this pattern - ![Contract Unit Test](tests/contract-testing.test.ts).
+
+
+![System Under Test (SUT)](img/contract_testing.png) 
+
+### Goal
+
+Identify side effects of non-breaking schema changes on consumer applications.
+
+### Description
+
+Event driven architectures decouple producers and consumers at the infrastructure layer but they are still coupled at the application layer by the event contract. Consumers rely on the event contract and schema to write their processing logic. Contract testing uses sample events provided by producers to validate business logic on consumer side for strong guarantee of no breaking changes. Contract tests are owned by consumers. As an example, changing the format of an “address” field from full address to only street address and adding new event fields like country, city, state and postcode would be considered a backward compatible schema change, since it’s only adding new elements. However, a consumer test expecting the full address in the “address’ field will fail and hence we will know that this change is actually not backward compatible.
+
+There are other ways to implement contract testing. For instance, one of the common ways is to use test doubles on the consumer and check that all the calls to your test doubles return the same results as a call to the real application would. Review the Pact.io documentation for an example of this approach.
+
+
+### Limitations
+
+There is upfront effort required on consumer side to write tests and on-going maintenance efforts. However once written, the tests are reusable. It takes longer to roll out schema changes and increases collaboration overhead to resolve breaking changes. 
+
+Customers can be selective in using contract testing for critical applications where cost of breaking changes is much higher than the additional work of extra tests and collaboration. For applications with lax SLA, not in critical path, they can trade off strong guarantee in favour of lower overhead by using only schema testing.
+
+
+---
+
+### Key Files in this Project
+
+  - [schema-testing.test.ts](tests/schema-testing.test.ts) - unit test for schema testing
+  - [contract-testing.test.ts](tests/contract-testing.test.ts) - unit test for contract testing
+  - [schemas](schemas) - versioned schemas used for schema testing
+  - [events](events) - sample events used for contract testing
+
+## Run the Tests
```

Comment:
Thinking we include an "install dependencies" step here, what do you think?

e.g. `npm install`

Reply:
Added